addons/im_livechat/static/src/core/web/livechat_channel_command_provider.js

```diff
@@ -0,0 +1,47 @@
+import { Component } from "@odoo/owl";
+import { registry } from "@web/core/registry";
+import { LivechatChannel } from "@im_livechat/core/web/livechat_channel_model";
+
+class LivechatChannelCommand extends Component {
+    static template = "im_livechat.LivechatChannelCommand";
+    static props = {
+        channel: LivechatChannel,
+        executeCommand: Function,
+        name: String,
+        searchValue: String,
+        slots: Object,
+    };
+}
+
+registry.category("command_provider").add("im_livechat.channel_join_leave", {
+    /**
+     * @param {import("@web/env").OdooEnv} env
+     */
+    async provide(env) {
+        const store = env.services["mail.store"];
+        if (!store?.hasLivechatAccess) {
+            return [];
+        }
+        await store.livechatChannels.fetch();
+        const activeChannels = new Set(
+            Object.values(store.LivechatChannel.records)
+                .filter((c) => c.threads.length > 0)
+                .map((c) => c.id)
+        );
+        // Show live chat channels with ongoing conversations first
+        return Object.values(store.LivechatChannel.records)
+            .sort((c) => (activeChannels.has(c.id) ? -1 : 1))
+            .map((c) => ({
+                async action() {
+                    if (c.hasSelfAsMember) {
+                        await c.leave();
+                    } else {
+                        await c.join();
+                    }
+                },
```

Comment:
Isn't this possibly confusing? The action name is based on the state of the channel when the list is rendered, but the action effect will be based on the state when the action is selected, which can be the opposite.

Reply:
I don't get your point: the action name will be updated re-actively according to `channel.hasSelfAsMember`. Let say the state changes after the command palette opening: the command palette item will be updated according to the new value thus will be in line with the action?

Reply:
Ok then it's probably fine, but I'm confused because name is given as props so this one will not be reactive. So I guess we have the same "problem" if the channel gets renamed.

Reply:
Updated not to pass name as props: name is now computed in the component itself based on `props.channel`.